mismi-cli/main/s3.hs

```diff
@@ -125,26 +142,113 @@ run c = do
     Write a t w ->
       writeWithModeOrFail w a t
     Read a ->
-      read a >>= \md -> liftIO $ maybe exitFailure pure md >>= putStrLn
+      read a >>= \md -> liftIO $ maybe exitFailure pure md >>= T.putStrLn
     Cat a ->
       read' a >>= \md -> liftIO $ maybe exitFailure pure md >>= runResourceT . ($$+- DC.mapM_ (liftIO . BS.putStr))
-    Size a ->
-      getSize a >>= liftIO . maybe exitFailure (putStrLn . pack . show)
+    Size a NotRecursive _ u ->
+      getSize a >>= liftIO . maybe exitFailure (T.putStrLn . renderSize u . Bytes . fromIntegral)
+    Size a Recursive d u ->
+      sizeRecursive a d u
     Sync s d m f ->
       renderExit renderSyncError $ syncWithMode m s d f
     List a rq ->
-      rec (list' a) (listRecursively' a) rq $$ DC.mapM_ (liftIO . putStrLn . addressToText)
+      rec (list' a) (listRecursively' a) rq $$ DC.mapM_ (liftIO . T.putStrLn . addressToText)
+
+sizeRecursive :: Address -> Detail -> UnitPrefix -> AWS ()
+sizeRecursive root d p =
+  case d of
+    All ->
+      getSizeRecursively' root $$ DC.mapM_ (liftIO . T.putStrLn . renderSizedAddress p)
+    Summary -> do
+      bytes <- getSizeRecursively' root $$ DC.map sizedBytes =$= DC.fold (+) 0
+      liftIO . T.putStrLn $ renderSizedAddress p (Sized bytes root)
+
+renderSizedAddress :: UnitPrefix -> Sized Address -> Text
+renderSizedAddress p (Sized bytes address) =
+  let
+    separator =
+      case p of
+        NoPrefix ->
+          "\t"
+        IEC ->
+          "  "
+        Metric ->
+          "  "
+  in
+    renderSize p bytes <> separator <> addressToText address
+
+renderSize :: UnitPrefix -> Bytes -> Text
+renderSize p bytes =
+  case p of
+    NoPrefix ->
+      T.pack (show $ unBytes bytes)
+    IEC ->
+      renderUnit iec bytes
+    Metric ->
+      renderUnit metric bytes
+
+iec :: Map Bytes Text
+iec =
+  Map.fromList [
+      (pow 1024 0, "B  ")
+    , (pow 1024 1, "KiB")
+    , (pow 1024 2, "MiB")
+    , (pow 1024 3, "GiB")
+    , (pow 1024 4, "TiB")
+    , (pow 1024 5, "PiB")
+    , (pow 1024 6, "EiB")
+    ]
+
+metric :: Map Bytes Text
+metric =
+  Map.fromList [
+      (pow 1000 0, "B ")
+    , (pow 1000 1, "kB")
+    , (pow 1000 2, "MB")
+    , (pow 1000 3, "GB")
+    , (pow 1000 4, "TB")
+    , (pow 1000 5, "PB")
+    , (pow 1000 6, "EB")
+    ]
+
+pow :: Bytes -> Int64 -> Bytes
+pow b p =
+  b ^ p
+
+renderUnit :: Map Bytes Text -> Bytes -> Text
+renderUnit prefixes n0 =
+  let
+    minView =
+      fromMaybe
+        (Savage.error "s3.renderUnit: empty map")
```

Comment:
I'm now keeping the zero element outside the map so we don't need to use a partial function